data_model/src/metadata.rs

```diff
@@ -1,27 +1,92 @@
 //! Metadata: key-value pairs that can be attached to accounts, transactions and assets.
 
 #[cfg(not(feature = "std"))]
-use alloc::{collections::btree_map, format, string::String, vec::Vec};
+use alloc::{
+    collections::btree_map,
+    format,
+    string::{String, ToString},
+    vec::Vec,
+};
 use core::borrow::Borrow;
 #[cfg(feature = "std")]
 use std::collections::btree_map;
 
 use derive_more::Display;
 use iroha_data_model_derive::model;
+use iroha_macro::FromVariant;
+use iroha_primitives::numeric::Numeric;
 use iroha_schema::IntoSchema;
 use parity_scale_codec::{Decode, Encode};
 use serde::{Deserialize, Serialize};
 
 pub use self::model::*;
-use crate::{Name, Value};
+use crate::Name;
+
+/// A path slice, composed of [`Name`]s.
+pub type Path = [Name];
 
 /// Collection of parameters by their names.
-pub type UnlimitedMetadata = btree_map::BTreeMap<Name, Value>;
+pub type UnlimitedMetadata = btree_map::BTreeMap<Name, MetadataValueBox>;
 
 #[model]
 pub mod model {
     use super::*;
 
+    /// Collection of parameters by their names with checked insertion.
+    #[derive(
+        Debug,
+        Display,
+        Clone,
+        Default,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Deserialize,
+        Serialize,
+        Decode,
+        Encode,
+        IntoSchema,
+    )]
+    #[ffi_type(opaque)]
+    #[repr(transparent)]
+    #[serde(transparent)]
+    #[display(fmt = "Metadata")]
+    #[allow(clippy::multiple_inherent_impl)]
+    pub struct Metadata(pub(super) btree_map::BTreeMap<Name, MetadataValueBox>);
+
+    /// Metadata value
+    #[derive(
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        FromVariant,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    #[ffi_type(opaque)]
+    #[allow(missing_docs)]
+    pub enum MetadataValueBox {
+        Bool(bool),
+        String(String),
+        Name(Name),
+        Bytes(Vec<u8>),
+        Numeric(Numeric),
+        LimitedMetadata(Metadata),
+
+        Vec(
+            #[skip_from]
+            #[skip_try_from]
+            Vec<MetadataValueBox>,
+        ),
+    }
```

Comment:
I think the values in metadata should be reconsidered. I'd be in favor of loosely typing it with `StringWithJson` and defer the parsing of the structure to executor as we do for permission tokens payload

Reply:
#4353